### DISCUSSION TITLE
FluidSynth source make error : 'KSDATAFORMAT_SUBTYPE_IEEE_FLOAT' undeclared (first use in this function)

### DISCUSSION THREAD

**Discussion by nijian141893** [FluidSynth/fluidsynth#1124] [Q&A]
$ mingw32-make
[ 14%] Built target gentables
[ 16%] Building C object src/CMakeFiles/libfluidsynth-OBJ.dir/drivers/fluid_waveout.c.obj
C:\Users\nijian02\Downloads\fluidsynthBuild\fluidsynth-2.2.7\src\drivers\fluid_waveout.c: In function 'new_fluid_waveout_audio_driver2':
C:\Users\nijian02\Downloads\fluidsynthBuild\fluidsynth-2.2.7\src\drivers\fluid_waveout.c:296:25: error: 'KSDATAFORMAT_SUBTYPE_IEEE_FLOAT' undeclared (first use in this function)
         wfx.SubFormat = KSDATAFORMAT_SUBTYPE_IEEE_FLOAT;
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Users\nijian02\Downloads\fluidsynthBuild\fluidsynth-2.2.7\src\drivers\fluid_waveout.c:296:25: note: each undeclared identifier is reported only once for each function it appears in
C:\Users\nijian02\Downloads\fluidsynthBuild\fluidsynth-2.2.7\src\drivers\fluid_waveout.c:305:25: error: 'KSDATAFORMAT_SUBTYPE_PCM' undeclared (first use in this function)
         wfx.SubFormat = KSDATAFORMAT_SUBTYPE_PCM;
                         ^~~~~~~~~~~~~~~~~~~~~~~~
src\CMakeFiles\libfluidsynth-OBJ.dir\build.make:90: recipe for target 'src/CMakeFiles/libfluidsynth-OBJ.dir/drivers/fluid_waveout.c.obj' failed
mingw32-make[2]: *** [src/CMakeFiles/libfluidsynth-OBJ.dir/drivers/fluid_waveout.c.obj] Error 1
CMakeFiles\Makefile2:182: recipe for target 'src/CMakeFiles/libfluidsynth-OBJ.dir/all' failed
mingw32-make[1]: *** [src/CMakeFiles/libfluidsynth-OBJ.dir/all] Error 2
Makefile:164: recipe for target 'all' failed
mingw32-make: *** [all] Error 2

how to solve this error, please help me.

**Comment by ghost** [comment]
是不是你的 MinGW 编译器太老了，建议使用 MSYS2 里的 MinGW64 才是最新的

**Reply by nijian141893** [reply]
> 是不是你的 MinGW 编译器太老了，建议使用 MSYS2 里的 MinGW64 才是最新的

我也不清楚为啥，编译fluidsynth源码太痛苦了...

请问，你有成功的编译经验吗

**Reply by ghost** [reply]
我用Clang64编译成功了，而且是单DLL文件的，调用它时无需再加载其他DLL。
二进制在我的fork里：
https://github.com/3119369616/fluidsynth/releases/tag/v2.2.7

**Reply by nijian141893** [reply]
> 我用Clang64编译成功了，而且是单DLL文件的，调用它时无需再加载其他DLL。 二进制在我的fork里： https://github.com/3119369616/fluidsynth/releases/tag/v2.2.7

嗯嗯，感谢。我是做安卓开发的，我需要用fluidsynth的so文件，因为我想改动下源码，所以需要自己编译。后来我跟着一个博客的介绍，在linux环境下编译出来安卓可以使用的so文件了。博客地址：https://www.codeleading.com/article/49755755328/
。现在我在阅读fluidsynth的源码，好难啊，看不懂...
